File: exercises/domain-1/block-1.1/solution/solution.py
import anthropic
import json

client = anthropic.Anthropic()
MAX_ITERATIONS = 5


def read_file(path: str) -> str:
    return f"""# Simulated file content for {path}
def add(a, b):
    return a + b

def divide(a, b):
    return a / b  # BUG: no zero check
"""


def analyze_code(code: str) -> str:
    if "divide" in code and "b == 0" not in code:
        return "Issue found: divide() has no zero-division guard. Add: if b == 0: raise ValueError('Division by zero')"
    return "No issues found — code looks clean"


def process_tool(name: str, tool_input: dict) -> str:
    if name == "read_file":
        return read_file(tool_input["path"])
    elif name == "analyze_code":
        return analyze_code(tool_input["code"])
    return f"Unknown tool: {name}"


def run_code_review_agent(file_path: str) -> str:
    tools = [
        {"name": "read_file", "description": "Read file contents",
         "input_schema": {"type": "object", "properties": {"path": {"type": "string"}}, "required": ["path"]}},
        {"name": "analyze_code", "description": "Analyze code for issues",
         "input_schema": {"type": "object", "properties": {"code": {"type": "string"}}, "required": ["code"]}},
    ]

    messages = [{"role": "user", "content": f"Review the code in {file_path} for issues. Use read_file first, then analyze_code."}]

    iterations = 0
    final_response = ""

    while iterations < MAX_ITERATIONS:
        response = client.messages.create(
            model="claude-sonnet-4-6",
            max_tokens=1024,
            tools=tools,
            messages=messages,
        )
        iterations += 1

        if response.stop_reason == "end_turn":
            for block in response.content:
                if block.type == "text":
                    final_response = block.text
            break

        if response.stop_reason == "tool_use":
            messages.append({"role": "assistant", "content": response.content})

            tool_results = []
            for block in response.content:
                if block.type == "tool_use":
                    result = process_tool(block.name, block.input)
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block.id,
                        "content": result,
                    })

            messages.append({"role": "user", "content": tool_results})

    if iterations >= MAX_ITERATIONS and not final_response:
        final_response = "Max iterations reached — partial analysis completed"

    return final_response


if __name__ == "__main__":
    print(run_code_review_agent("src/math.py"))

File: exercises/domain-1/block-1.1/solution/solution.ts
import Anthropic from "@anthropic-ai/sdk"

const client = new Anthropic()
const MAX_ITERATIONS = 5

function readFile(path: string): string {
  return `// Simulated file content for ${path}\nfunction add(a, b) { return a + b }\nfunction divide(a, b) { return a / b }  // BUG: no zero check`
}

function analyzeCode(code: string): string {
  if (code.includes("divide") && !code.includes("b === 0")) {
    return "Issue found: divide() has no zero-division guard. Add: if (b === 0) throw new Error('Division by zero')"
  }
  return "No issues found — code looks clean"
}

function processTool(name: string, input: Record<string, string>): string {
  switch (name) {
    case "read_file": return readFile(input.path)
    case "analyze_code": return analyzeCode(input.code)
    default: return `Unknown tool: ${name}`
  }
}

async function runCodeReviewAgent(filePath: string): Promise<string> {
  const tools: Anthropic.Tool[] = [
    {
      name: "read_file",
      description: "Read the contents of a file",
      input_schema: { type: "object" as const, properties: { path: { type: "string" } }, required: ["path"] },
    },
    {
      name: "analyze_code",
      description: "Analyze code for issues",
      input_schema: { type: "object" as const, properties: { code: { type: "string" } }, required: ["code"] },
    },
  ]

  const messages: Anthropic.MessageParam[] = [
    { role: "user", content: `Review the code in ${filePath} for issues. Use read_file first, then analyze_code.` },
  ]

  let iterations = 0
  let finalResponse = ""

  while (iterations < MAX_ITERATIONS) {
    const response = await client.messages.create({
      model: "claude-sonnet-4-6",
      max_tokens: 1024,
      tools,
      messages,
    })

    iterations++

    if (response.stop_reason === "end_turn") {
      // Extract final text
      const textBlock = response.content.find((b) => b.type === "text")
      finalResponse = textBlock?.type === "text" ? textBlock.text : "No response"
      break
    }

    if (response.stop_reason === "tool_use") {
      // Add assistant message
      messages.push({ role: "assistant", content: response.content })

      // Process all tool calls and build tool_result blocks
      const toolResults: Anthropic.ToolResultBlockParam[] = response.content
        .filter((b): b is Anthropic.ToolUseBlock => b.type === "tool_use")
        .map((toolUse) => ({
          type: "tool_result" as const,
          tool_use_id: toolUse.id,
          content: processTool(toolUse.name, toolUse.input as Record<string, string>),
        }))

      messages.push({ role: "user", content: toolResults })
    }
  }

  if (iterations >= MAX_ITERATIONS && !finalResponse) {
    finalResponse = "Max iterations reached — partial analysis completed"
  }

  return finalResponse
}

runCodeReviewAgent("src/math.ts").then(console.log).catch(console.error)
